ext/pdo/pdo_stmt.c

```diff
@@ -840,18 +818,17 @@ static bool do_fetch(pdo_stmt_t *stmt, zval *return_value, enum pdo_fetch_type h
 					}
 				}
 				if (ce->constructor && (flags & PDO_FETCH_PROPS_LATE)) {
+					zval retval_constructor_call;
+					stmt->fetch.cls.fci.retval = &retval_constructor_call;
 					stmt->fetch.cls.fci.object = Z_OBJ_P(return_value);
 					stmt->fetch.cls.fcc.object = Z_OBJ_P(return_value);
 					if (zend_call_function(&stmt->fetch.cls.fci, &stmt->fetch.cls.fcc) == FAILURE) {
 						/* TODO Error? */
 						pdo_raise_impl_error(stmt->dbh, stmt, "HY000", "could not call class constructor");
-						return 0;
-					} else {
-						if (!Z_ISUNDEF(stmt->fetch.cls.retval)) {
-							zval_ptr_dtor(&stmt->fetch.cls.retval);
-							ZVAL_UNDEF(&stmt->fetch.cls.retval);
-						}
+						return false;
 					}
+					zval_ptr_dtor(&retval_constructor_call);
```

Comment:
Why this doesn't need the `Z_ISUNDEF` check anymore?

Reply:
IS_UNDEF only happens if there is an exception that is thrown by the constructor.
Actually the thing that is pointless is to check for `FAILURE` as the result of `zend_call_function()` as that can never happen as we are not in shutdown.